File: 0234-palindrome-linked-list/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode slow =head;
        ListNode fast = head;
        while(fast!=null && fast.next!=null){
            slow=slow.next;
            fast=fast.next.next;
        }
        ListNode first = head;
        ListNode sec = reverse(slow);
        while(sec!=null){
            if(first.val!=sec.val){
                return false;
            }
            first=first.next;
            sec=sec.next;
        }
        return true;
    }
    public ListNode reverse(ListNode head){
        ListNode prev = null;
        ListNode cur =head;
        while(cur!=null){
            ListNode newnode = cur.next;
            cur.next = prev;
            prev=cur;
            cur=newnode;
        }
        return prev;
    }
}

File: 0234-palindrome-linked-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        vector<int> arr;
        struct ListNode* temp =head;
        while(temp!=NULL){
            arr.push_back(temp->val);
            temp=temp->next;
        } 
        int l=0;
        int r=arr.size()-1;
        while(l<=r){
            if(arr[l]!=arr[r]){
                return false;
            }
            l++;
            r--;
        }
        return true;
    }
};
